### DISCUSSION TITLE
How can I access the current repo context and files from a Docker container action ?

### DISCUSSION THREAD

**Discussion by obrassard** [community/community#26855] [Actions]
<p>Hello ! I am trying to build a new Docker container action that must interact with the current repo files.<br>
I understand how to setup and initialize my action with Docker, however I do not understand how am i suppose to acces the repository’s files and the current job context Indeed, lets supose that we have a workflow that calls actions/checkout, how can I then acces these files from inside the container ? Do I have to mount a docker volume to GITHUB_WORKSPACE, or does Github Actions provide a better way to do this ?</p>
<p>Thanks in advance !</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p><a class="mention" href="/obrassard">@obrassard</a> When use a Docker container action in the workflow, it will automatically run with a volume mapping from GITHUB_WORKSPACE (on the runner machine) to /github/workspace (inside container). </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/de456053b4e214f711b8e2e465f21809e07887ca.png" data-download-href="/uploads/short-url/vIiDsqF4FWjJGyaJ85FGceSTI4W.png?dl=1" title="docker container actions.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181100297-ae085d0b-a4d1-4f35-a4dd-68fabeb17a58.png" alt="docker container actions.png" data-base62-sha1="vIiDsqF4FWjJGyaJ85FGceSTI4W" width="690" height="291" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/de456053b4e214f711b8e2e465f21809e07887ca_2_690x291.png, https://aws1.discourse-cdn.com/github/optimized/2X/d/de456053b4e214f711b8e2e465f21809e07887ca_2_1035x436.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/d/de456053b4e214f711b8e2e465f21809e07887ca_2_1380x582.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/de456053b4e214f711b8e2e465f21809e07887ca_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">docker container actions.png</span><span class="informations">1529×645 36.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>So you could easily get access to your repo files under / <strong>github/workspace</strong> inside the container.</p>
<p>For example, I want to access to the README.md in my repo,</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/7fe4c9c6e3f41bc955d7fae23ddd25c771f1d4b7.png" data-download-href="/uploads/short-url/ifoOxzArIeGOn0tGjJQ5Q5YJ91t.png?dl=1" title="readme.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181100299-b96a9dbe-5241-4ada-91ea-04d3406a78f4.png" alt="readme.png" data-base62-sha1="ifoOxzArIeGOn0tGjJQ5Q5YJ91t" width="636" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/7fe4c9c6e3f41bc955d7fae23ddd25c771f1d4b7_2_636x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/7/7fe4c9c6e3f41bc955d7fae23ddd25c771f1d4b7_2_954x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/7fe4c9c6e3f41bc955d7fae23ddd25c771f1d4b7.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/7fe4c9c6e3f41bc955d7fae23ddd25c771f1d4b7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">readme.png</span><span class="informations">970×762 45.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>In my Docker container action &gt; entrypoint.sh, I could use </p>
<pre><code>cat /github/workspace/README.md
</code></pre>
<p>to show the content of README.md. </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181100301-ca18a71c-358f-45fc-85d7-1192ae2c0770.png" alt="cat file.png" data-base62-sha1="khDiDOW66jR4HphOuLmRydMJPH3" width="597" height="334"></p>

**Reply by Dark-Kernel** [reply]
So like this we can access created files in container( generated with script) ,by moving it to /GitHub/workspace?
Suppose I one file is created when I run my script in container which is part of my repo
After completion of script it generates 1 file. Now I want that file  then do I move it to my repo of I can use it as it is? Will it be their for long time. ?

**Comment by obrassard** [comment]
<p>Thanks for your great answer !!!</p>
<p>I think this information should be better explained in the GitHub actions documentation. Have a nice day :+1:</p>

**Comment by Arrow7000** [comment]
<p>That doesn’t seem to be the case for me. I’m running the <code>checkout@v2</code> action before my Docker action but there doesn’t seem to be a <code>/github/workspace</code> directory containing my source files.</p>
<p>In my logs you can see that even though I’m trying to copy my <code>content</code> folder from <code>/github/workspace/content</code> it doesn’t seem to exist.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/030adf41485cfd2f98cab4e720ef16839cc1a7a5.png" data-download-href="/uploads/short-url/qUIZO5QPWDDemKPFYojUfpASUJ.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/13225609/181100306-d4cd738b-9217-4495-b95c-65e4852e8b47.png" alt="image" data-base62-sha1="qUIZO5QPWDDemKPFYojUfpASUJ" width="690" height="88" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/030adf41485cfd2f98cab4e720ef16839cc1a7a5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">823×106 5.74 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>What am I doing wrong?</p>

**Comment by airtower-luna** [comment]
<p>That’s part of the container <em>build</em>, not run. The workspace directory is only mounted at run time. Also Docker builds in general can’t access random files, only files in their build context (see <a href="https://docs.docker.com/engine/reference/builder/#usage">the Usage section of the Dockerfile reference</a>). Simply put, you can’t use files from the repository running your action in the Dockerfile, only later in the code that’s running in the resulting container.</p>

**Comment by obrassard** [comment]
<p>Indeed, you won’t have access to <code>/github/workspace</code> during the build phase of your docker image.<br>
This directory is attached (mounted as a docker volume) to your container instance once it is up and running. So, in other words, you will have access to <code>/github/workspace</code> when the action run, not when the image is built. To perform operations on the files in <code>/github/workspace</code>, you could set a shell script as your action entrypoint.</p>

**Comment by Arrow7000** [comment]
<p>Thanks for your quick reply!</p>
<p>I’ll be honest I don’t fully understand what the difference between the build and run phase is… Could you give me an example of how or at what point I <em>would</em> be able to use my repo files in a docker action?</p>
<p>Or if I can never access to my repo files in a docker action then I don’t really understand why anyone would ever choose docker for their action type. Surely the point in having an action is for it to be able to do something with your source code?</p>
<p>If there is documentation answering this question then please point me towards it, but I’m struggling to understand how actions work in anything more complicated than the basic examples.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="Arrow7000" data-post="7" data-topic="17711">
<div class="title">
<div class="quote-controls"></div>
 Arrow7000:</div>
<blockquote>
<p>If there is documentation answering this question then please point me towards it, but I’m struggling to understand how actions work in anything more complicated than the basic examples.</p>
</blockquote>
</aside>
<p>The distinction between building a container image and running it is a general Docker (and container) concept, so I’d start with the Docker introduction:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/5814919/181100308-ceb1df5b-f3b2-4f79-bc63-4cc1eda9fc7e.png" class="site-icon" width="129" height="128">
      <a href="https://docs.docker.com/get-started/" target="_blank" rel="noopener" title="02:16PM - 23 September 2020">Docker Documentation – 23 Sep 20</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:129/128;"><img src="https://user-images.githubusercontent.com/5814919/181100309-2b4445c9-d944-4843-af70-97a748d71b35.png" class="thumbnail" width="129" height="128"></div>

<h3><a href="https://docs.docker.com/get-started/" target="_blank" rel="noopener">Orientation and setup</a></h3>

<p>Orientation and setup Build and run your image Share images on Docker Hub Welcome! We are excited that you want to learn Docker. This page contains step-by-step instructions on how...</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Highly simplified: The Dockerfile describes what should be included in the container image, usually software and maybe supporting files. When the container gets <em>run</em>, the command you set as <code>ENTRYPOINT</code> in the Dockerfile is executed. In Github Actions that command will have access to the build data as described above.</p>

**Comment by Arrow7000** [comment]
<p>Ok thank you for that. I’ll have a read through the docs.</p>
<p>But is there a <em>reason</em> why docker shouldn’t have access to the repo files during the build phase? I can’t see why this should be the case. Why not make them available during the build phase too?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="Arrow7000" data-post="9" data-topic="17711">
<div class="title">
<div class="quote-controls"></div>
 Arrow7000:</div>
<blockquote>
<p>Why not make them available during the build phase too?</p>
</blockquote>
</aside>
<p>Because the build is about getting your action code ready to run. You can roughly compare it to compiling a piece of software, and depending on the action code there may be actual compiling involved. Needing the data a user might want to process while compiling a tool doesn’t make sense.</p>
<p>It’d also cause a bunch of technical issues, in particular how to make which files available to the Docker build. That should become clear when you read about the <em>build context</em> in the Docker documentation.</p>

**Reply by nemith** [reply]
The `docker build` command at least allows settings the context and having it separate from the Dockerfile.

**Comment by Arrow7000** [comment]
<p>Ok I think that makes sense. I’ll have a look at the docs like you recommended.</p>
<p>Thanks again for your help!</p>

**Comment by oshliaer** [comment]
This might also be very useful https://github.com/orgs/community/discussions/26123
